src/test/java/DirectionsServletTest.java

```diff
@@ -0,0 +1,145 @@
+  private static final String MONTREAL_TO_MONTREAL_WITH_WAYPOINTS_JSON =
+      "[\"[DirectionsLeg: \"Montreal, QC, Canada\" -\u003e \"Kitchener, ON, Canada\" (45.50171230,-73.56721840 -\u003e 43.45184990,-80.49313410), duration\u003d6 hours 10 mins, distance\u003d632 km: 16 steps]\",\"[DirectionsLeg: \"Kitchener, ON, Canada\" -\u003e \"Waterloo, ON, Canada\" (43.45184990,-80.49313410 -\u003e 43.46430180,-80.52042120), duration\u003d8 mins, distance\u003d3.0 km: 5 steps]\",\"[DirectionsLeg: \"Waterloo, ON, Canada\" -\u003e \"Windsor, ON, Canada\" (43.46430180,-80.52042120 -\u003e 42.31486680,-83.03656800), duration\u003d2 hours 55 mins, distance\u003d291 km: 18 steps]\",\"[DirectionsLeg: \"Windsor, ON, Canada\" -\u003e \"Montreal, QC, Canada\" (42.31486680,-83.03656800 -\u003e 45.50171230,-73.56721840), duration\u003d8 hours 42 mins, distance\u003d897 km: 21 steps]\"]";
+  private static final String MONTREAL_TO_WATERLOO_NO_WAYPOINTS_JSON =
+      "[\"[DirectionsLeg: \"Montreal, QC, Canada\" -\u003e \"Waterloo, ON, Canada\" (45.50171230,-73.56721840 -\u003e 43.46430180,-80.52042120), duration\u003d6 hours 11 mins, distance\u003d638 km: 22 steps]\"]";
+
+  @Before
+  public void setUp() throws IOException {
+    request = Mockito.mock(HttpServletRequest.class);
+    response = Mockito.mock(HttpServletResponse.class);
+    directionsClientFactory = Mockito.mock(DirectionsClientFactory.class);
+    directionsClient = Mockito.mock(DirectionsClient.class);
+    servlet = new DirectionsServlet(directionsClientFactory);
+    // Writer used in get/post requests to capture HTTP response values
+    stringWriter = new StringWriter();
+    printWriter = new PrintWriter(stringWriter);
+    Mockito.when(directionsClientFactory.getDirectionsClient(Mockito.any()))
+        .thenReturn(directionsClient);
+    Mockito.when(response.getWriter()).thenReturn(printWriter);
+  }
+
+  @After
+  public void tearDown() throws IOException {
+    // Dump contents after each test
+    stringWriter.getBuffer().setLength(0);
+    stringWriter.close();
+    printWriter.close();
+  }
+
+  @Test
+  public void windsorToMontrealWithWaypoints() throws DirectionsException, ServletException {
+    // Get optimized route of travel from Windsor, ON to Montreal, QC with waypoints
+    // Markham, ON,
+    // Waterloo, ON and Quebec City, QC in between.
+    // Should return four legs: Windsor, ON -> Waterloo, ON -> Markham, ON -> Quebec
+    // City, QC ->
+    // Montreal, QC.
+    Mockito.when(
+            directionsClient.getDirections(
+                Mockito.anyString(), Mockito.anyString(), Mockito.anyListOf(String.class)))
+        .thenReturn(WINDSOR_TO_MONTREAL_WITH_WAYPOINTS);
+    servlet.doGet(request, response);
+    printWriter.flush();
```

Comment:
Food for thought: every test has to flush the `PrintWriter`, and this is error-prone. Consider adding an adapter around `PrintWriter` that flushes on every print automatically. (It's outside the scope of this CL.)

Reply:
This is interesting, i've created an issue for this #94.

Reply:
I think you can now use the `HttpServletResponseFake` class.

Reply:
Added, thanks!

Reply:
`RESOLVED`